### PR TITLE
Issue-20619 early return existing value.

### LINKED ISSUE #20619: Early return possible in ImmutableInterlocked.AddOrUpdate
If the `valueFactory` returns the same value as was present in the dictionary previously, the [`AddOrUpdate`](https://github.com/dotnet/corefx/blob/343561048c85321713c30b6e69ddd20cfac94dea/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs#L262) method can return that immediately. This optimization would be equivalent to the one currently performed by the [`Update`](https://github.com/dotnet/corefx/blob/343561048c85321713c30b6e69ddd20cfac94dea/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs#L42-L46) methods in the same class.

:memo: There are two `AddOrUpdate` methods which this change applies to.

### PR DESCRIPTION
**Pull request by lateapexearlyspeed**
Fix #20619

### CODE REVIEW

**Comment by eiriktsarpalis** [comment]
Changes look good, but we should probably address [this issue](https://github.com/dotnet/runtime/issues/20619#issuecomment-754497096) before merging.

**Comment by eiriktsarpalis** [comment]
Hi @lateapexearlyspeed, any updates concerning the feedback on this PR?

**Comment by lateapexearlyspeed** [comment]
Hi @eiriktsarpalis, I am waiting for feedback. Hi @sharwell, any thought for linked question above?

**Comment by eiriktsarpalis** [comment]
I would say we should just go ahead and replace the hardcoded Comparer<T>.Default calls, since they're clearly omissions.

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs
Comment (sharwell) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs:350]: 💭 It looks like `SetItem` already performs a value check. If you are looking for a way to ensure this method does not end up performing more [CAS](https://en.wikipedia.org/wiki/Compare-and-swap) operations than necessary, we could do a reference equality check between `updatedCollection` and `priorCollection` between the call to `SetItem` and the call to `CompareExchange`. This also avoids any issues with the use of a specific value comparer and virtual/interface dispatch overhead it could incur.

Comment (lateapexearlyspeed) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs]: @sharwell Fixed, thanks.

**Comment by lateapexearlyspeed** [comment]
> I would say we should just go ahead and replace the hardcoded `Comparer<T>.Default` calls, since they're clearly omissions.

@eiriktsarpalis Fixed, thanks.

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs
Comment (sharwell) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs:463]: 📝 Given we are no longer using value comparers as part of the new change, perhaps we submit this as a separate PR?

Comment (eiriktsarpalis) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs:463]: > Given we are no longer using value comparers as part of the new change

What change are you referring to?

Comment (sharwell) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs:463]: https://github.com/dotnet/runtime/pull/46481#discussion_r566358634

Comment (sharwell) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs:463]: I'm suggesting we revert the changes to `TryUpdate` from this pull request.

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs]: whitespace issues seem to be failing the build.

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs:463]: Ah, I missed that change. Yeah I think that's reasonable.

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs
Comment (lateapexearlyspeed) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs]: Already fixed, this should be a 'Outdated' file change.

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs
Comment (lateapexearlyspeed) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs]: Fixed, thanks.

**Review by eiriktsarpalis — APPROVED** [review]

**Comment by eiriktsarpalis** [comment]
Thank you for your contribution!

**Pull request merged**
